Issue/2291 print shipping labels

anitaa1990 opened the pull request:
Fixes #2291 by adding the option to print a shipping label.

#### Changes
- Adds new fragment/module/viewmodel to print a shipping label.
- Adds a dialog to display the different paper sizes when printing a shipping label.

#### Screenshots
<img src="https://user-images.githubusercontent.com/22608780/94119742-f21aff00-fe6c-11ea-92f0-10a5e26b0fa4.gif" width="300"/>

#### Testing
- Click on an order from the Orders tab that has shipping labels.
- Notice that if the shipping label is refunded, the `Print shipping label` button is NOT displayed.
- Click on the `Reprint shipping label` button.
- Click on the paper size text field and choose any paper size.
- Click on the `Print shipping label button`.
- Notice the progress bar displayed. A pdf should be opened with a preview of the shipping label.

**Note that both the info options in the Print shipping label screen is not clickable right now but will be taken care of in another issue in #2295** 

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

anitaa1990 committed:
Add button to shipping label item to print the label

anitaa1990 committed:
Add click listener to print shipping label button

anitaa1990 committed:
Added logic to redirect to print shipping label fragment on button click

anitaa1990 committed:
Add method to repository to print shipping label

anitaa1990 committed:
Added UI to print shipping label

anitaa1990 committed:
Added logic to display paper size selector dialog when paper size field is clicked

anitaa1990 committed:
Added logic to display the shipping label preview for printing

Garance91540 reviewed (dismissed):
👏

anitaa1990 committed:
Merge branch 'develop' into issue/2291-print-shipping-labels

# Conflicts:
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/OrderNavigationTarget.kt
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/details/OrderDetailFragment.kt
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/details/adapter/OrderDetailShippingLabelsAdapter.kt
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/details/views/OrderDetailShippingLabelsView.kt
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/ShippingLabelRefundFragment.kt
#	WooCommerce/src/main/res/values/strings.xml

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/media/FileUtils.kt:
```diff
@@ -0,0 +1,39 @@
+    fun createTempPdfFile(storageDir: File): File {
+        val timeStamp: String = SimpleDateFormat("yyyyMMdd_HHmmss", Locale.getDefault()).format(Date())
+        val imageFileName = "PDF_" + timeStamp + "_"
+        return File.createTempFile(imageFileName, ".pdf", storageDir)
```
Note that `createTempFile` may throw either an `IOException` or a `SecurityException`, so I think we want to catch those here and return null, and update the calling function to show an error message in this situation.

anitaa1990 replied:
👍 Fixed in 4ab33e5

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/media/FileUtils.kt:
```diff
@@ -0,0 +1,39 @@
+    /**
+     * Creates a temp pdf file to store the shipping label pdf
+     */
+    fun createTempPdfFile(storageDir: File): File {
```
I'm wondering if we should make this more generic by adding a `fileExtension` parameter? That way we could use this for any type of file, not just a PDF.

anitaa1990 replied:
Good call! Fixed in acb067b

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/media/FileUtils.kt:
```diff
@@ -0,0 +1,39 @@
+    /**
+     * writes the preview string for a shipping label into the [tempFile]
+     */
+    fun writePdfToTempFile(tempFile: File, stringToWrite: String): File? {
```
Same comment here, should this be more generic? Right now it's only used for shipping labels, but it may be used for other files down the road.

anitaa1990 replied:
👍 Fixed in acb067b

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/media/FileUtils.kt:
```diff
@@ -0,0 +1,39 @@
+
+            tempFile
+        } catch (e: java.lang.Exception) {
+            e.printStackTrace()
```
I think we'd want to log this exception, rather than just printing the stack trace.

anitaa1990 replied:
👍 Fixed in 4ab33e5

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/PrintShippingLabelFragment.kt:
```diff
@@ -0,0 +1,117 @@
+    }
+
+    private fun openShippingLabelPreview(base: String) {
+        requireContext().getExternalFilesDir(Environment.DIRECTORY_DOCUMENTS)?.let {
```
Do we want to display an error toast if either `getExternalFilesDir` or `writeShippingLabelToFile` fails (ie: returns `null`)?

anitaa1990 replied:
Good call! Fixed in 4ab33e5

nbradbury commented:
@anitaa1990 If I rotate the device while the paper size dialog is showing, I can no longer choose a different paper size.

![ezgif-2-79e091b6c5c2](https://user-images.githubusercontent.com/3903757/95076040-403ac880-06df-11eb-919b-5297b0511b60.gif)

nbradbury commented:
> Note that both the info options in the Print shipping label screen is not clickable right now but will be taken care of in another issue in #2295

Since this is being merged into `develop`, perhaps we should hide these until they're clickable?

nbradbury commented:
@Garance91540 The wording here seems unfriendly. Is there a nicer way we can say this?

![Screenshot_1601898815](https://user-images.githubusercontent.com/3903757/95076592-17ff9980-06e0-11eb-8688-d21ec805adb2.png)

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/ShippingLabelPaperSizeSelectorDialog.kt:
```diff
@@ -0,0 +1,53 @@
+
+        companion object {
+            fun fromString(type: String): ShippingLabelPaperSize {
+                return when (type.toLowerCase(Locale.US)) {
```
This function doesn't appear to be used, but if it is it will always return `LABEL` because `type` is being converted to lower case and all the `when` cases are upper case.

anitaa1990 replied:
Good catch! I removed the method in 26b0b9e 👍

nbradbury commented:
@anitaa1990 Done with my first round!

Garance91540 commented:
> @Garance91540 The wording here seems unfriendly. Is there a nicer way we can say this?

@nbradbury I'm using the messaging we have in core, and agree it's not super friendly. I think it's also part of the goal (so to prevent frauders), but we could de-emphasize it a bit to make it less agressive at least:

https://d.pr/i/OPkFUi

@anitaa1990 let me know if that works for you

anitaa1990 committed:
Removed redundant method from ShippingLabelPaperSizeSelectorDialog

anitaa1990 committed:
Updated disclaimer design after feedback

anitaa1990 committed:
Added error messages when preview shipping label fails

anitaa1990 committed:
Merge branch 'develop' into issue/2291-print-shipping-labels

anitaa1990 committed:
Fixed bug where the paper size was not updated after screen rotation

anitaa1990 committed:
Renamed methods in FileUtils to make it more generic

anitaa1990 commented:
Thanks for the thorough review @nbradbury 😍 

> @Garance91540 The wording here seems unfriendly. Is there a nicer way we can say this?
@nbradbury I'm using the messaging we have in core, and agree it's not super friendly. I think it's also part of the goal (so to prevent frauders), but we could de-emphasize it a bit to make it less agressive at least:

I've updated the wording based on @Garance91540 's changes. This is what it looks like now:

<img src="https://user-images.githubusercontent.com/22608780/95192039-4029ee00-07ef-11eb-900a-06e945f80c40.png" width="300"/>. <img src="https://user-images.githubusercontent.com/22608780/95192047-41f3b180-07ef-11eb-8179-2ba4abafdd7f.png" width="300"/>

> Since this is being merged into develop, perhaps we should hide these until they're clickable?

The shipping labels feature right now is only active for debug users so I think we're good to merge this PR as is. WDYT?

>  @anitaa1990 If I rotate the device while the paper size dialog is showing, I can no longer choose a different paper size.

Thanks to @0nko's help I was able to fix this in fbeb8c8! Nice catch!

I believe I've addressed all the issues @nbradbury! Ready for another round 👍

anitaa1990 committed:
Updated logs to include exception

anitaa1990 commented:
> np: Should we use WooLog.e here and in writePdfToTempFile and include the exception in the log? ie:

🤦‍♀️ I thought I did do that. My bad! Updated in ea93458. Thanks for noticing that!

nbradbury reviewed (approved):
:shipit:
